### DISCUSSION TITLE
Add an analog AUX input port and ADC to enable it acting as an music network streamer source?

### DISCUSSION THREAD

**Discussion by Hedda** [FutureProofHomes/Satellite1-Hardware#21] [Ideas]
@FutureProofHomes Would you please consider adding a line-level stereo audio port and ADC (analog-to-digital converter) for analogue audio input to I2S to a future ESP32 and ESPHome-based hardware device in your product series in order to enable that appliance acting as an analog-to-digital music network streamer source?

Basically asking for a specialized ESP32-based sound capture and ADC + audio encoder product, specifially optimized for one specific use case: being a very user-friendly appliance for digitizing and streaming analogue sources while having the best possible audio recording quality. 

Regardless if built-into future Satellite1 hardware or created as a a brand new separate stand-alone product running different ESPHome firmware, I believe that this concept would be a great value-add feature if could as an option also use these type of smart-speaker as simple music streamer appliance with analog AUX input to use as audio source provider for Music Assistant. 

- https://www.home-assistant.io/blog/2024/05/09/music-assistant-2
  - https://www.home-assistant.io/blog/2025/03/05/music-assistants-next-big-hit/

High-quality audio music streamers are becoming more common but most of them are relativly expensive and all of them are only work within their own walled-in ecosystem (usually only workibg within the same brand, and Sonos probably being the most well known brand). 

The fact is that having this feature is a large piece that the community is missing, there are no other easy-to-use product for this ability as a simple Music provider solutions for Music Assistant to stream audio from analog AUX input sources. Especially not any complete hardware running ESPHome firmware. 

Meaning there is a demand and should be a market for an easy way to achieve a remote AUX input into Music Assistant from an external analog audio source like a vinyl record player (LP turntable) or cassette player.

What the product should achieve with this feature request idea is to have solution with support in both Music Assistant server and ESPHome firmware a small ESP32-based device that has audio line-level ADC (Analogue-to-Digital Converter) for audio input from a connected analogue audio HiFi equipment like a Turntable, CD-Player or Cassette-Player, and for software in ESPHome on that device to stream that audio to Music Assistant which in turn will forward to any speaker group that Music Assistant supports.

My goal is to replace Darkice/Icecast based solution running on Raspberry Pi as linked here with a simple stand-alone ESP32-based WiFi device:

* https://www.music-assistant.io/faq/stream-to/#music-assistant
  * https://github.com/gieljnssns/darkice-libaacplus-rpi-guide/blob/master/README.md
  * https://maker.pro/raspberry-pi/projects/how-to-build-an-internet-radio-station-with-raspberry-pi-darkice-and-icecast
  * https://github.com/gieljnssns/darkice-libaacplus-rpi-guide/blob/master/README.md

See directly related feature request for Music Assistant:

- https://github.com/orgs/music-assistant/discussions/2343

What I would like to see a such product achieve is a basic solution that is easy to install/maintain and use that allow any less-than-tech-savvy-person to stream music from a vinyl record player (LP turntable) or analogue casette-player to any speaker or group of speakers in our home. The vinyl record player (turntable) setup she has a pre-amp with phono (RCA) output ports for analog audio in stereo.

I would therefore prefer if we could buy some kind of networked (Wi-Fi) enabled appliance like a music streamer with stereo AUX input port that it will use for on-the-fly perform analog-to-digital conversion (ADC) + encoding for streaming to a Music Provider inside Music Assistant.

* **Architecture example:** Analog audio source with preamp -> ADC network appliance  -> onboard conversion, encosibg and formatting > digital music stream over the local network -> Music Assistant -> Any compatible speakers/reciever on the same network.

I have yet found a commercial product that does this outside the walled garden of closed ecosystems. That is, there are commercial products that support this but they will only stream to their own branded speakers within their own closed ecosystems (e.g. Sonos, and Yamaha).

Are there any solutions that can do this but stream to a Music Provider inside Music Assistant so it can forward to any Player Provider types?

**UPDATE:** To clearify, I would prefer to not build/maintain a DIY project myself for this, I would instead like to buy a complete appliance product that just works out-of-the-box and can be used by anyone without knowing how it actually works, like magic! ;)

I think that the hardware should preferably have a high-quality ADC (analog-to-digital converter) for Hi-Fi music streams if possible. Even though I'm not an audiophile myself I believe having the final product list such capability in its specifications would make it apeal to a wider audience that might not buy it otherwise.

Normally you have to use an expensive propriatory Hi-Fi music streamer system to achive this, howverr there are also mid-range products like the WiiM Ultra, WiiM Pro Plus, WiiM Amp, and WiiM Amp Pro use Linkplay streaming technology that enable analog and digital audio input ports, but practially all of them will only stream within their own ecosystem so you need to buy both music streamers and stream receivers from the same brand for all your speakers to make them compatible.

* https://www.wiimhome.com/wiimultra/overview
* https://www.wiimhome.com/wiimpro/overview
* https://www.wiimhome.com/wiimamp/overview

![Image](https://github.com/user-attachments/assets/86f56e35-0614-4cf8-95e3-21de80f1b764)

Alternativly I guess getting input via [Bluetooth audio streaming ](https://www.bluetooth.com/learn-about-bluetooth/solutions/audio-streaming/)could work in pinch but I am not aware of any such devices that are not troublesome and Bluetooth is not a very great way for permanent streaming which can be WAF approved (as it is my wifi that wants to stream her LP-records). 

PS: I have previously used the ”Vinyl Cast” app on an Android phone/tablet (with analog input and ADC via a USB Audio Device) for casting to Chromecast Audio using Google Cast which was user-friendly but it was very buggy as ”Vinyl Cast” app the kept crashing + aesthetically the solution as a whole was not WAF approved + it kind of ruined the nostalgic retro vibe having to use the interface of Android phone/tablet each time, (nice with it though was that at least the initial release of the Vinyl Cast app included using Audio ACR to detect the song being played, but unfortunately this feature relied on a now discontinued third-party library/service so the feature was removed.)

* https://github.com/aschober/vinyl-cast

![Image](https://github.com/user-attachments/assets/58ff3d1e-31a8-403a-8a4e-02c47ec58821)

**Comment by Hedda** [comment]
Again, even if possible to achieve this today with a creative DIY setup, that is nor easy nor simple to maintain, nor a complete all-in-one product that you can buy. 

Music Assistant how-to page for this currently only have suggestions on something that you can DIY build yourself using various parts with custom software and DIY hardware running on a Raspberry Pi, what is not what I would like/request here.

https://music-assistant.io/faq/stream-to/#music-assistant

So while you could use [Darkcast](http://www.darkice.org/) to capture and [Icecast](https://www.icecast.org/) to build a solution that will digitize and stream audio from your analog audio equipment like a vinyl record player (turntable/phonograph/gramophone) as a web radio stream (URL) that you could add as a radio station in Music Assistant, for such a DIY project you need an advanced audio-capture and ADC (analogue-to-digital converter) setup and compexed configuration that provides audio-input and digitalization. See proof-of-concept examples:

https://github.com/gieljnssns/darkice-libaacplus-rpi-guide/blob/master/README.md

And

https://maker.pro/raspberry-pi/projects/how-to-build-an-internet-radio-station-with-raspberry-pi-darkice-and-icecast

Those suggested DIY solution is also messy as it needs a lot of parts when your record player does not have a built-in USB Audio Device with ADC but instead the turntable only has analog audio out (RCA stereo output with or without pre-amp).

Please try to think about the same kind of people that would rather buy a Home Assistent Green applance instead of installing Home Assistant on a Raspberry Pi, so no DIY solution that will need to be maintained differently and seperatly.

 Then you add the WAF factor to that and try to see that not everyone want to have a DIY solution containing many parts with a mess of connected wires right next to their pretty turntable in the living room as the would ruin the aesthetics look and nostalgic + retro feeling that is one of the main purposes for using a retro turntable with LP records.

**Comment by alextrical** [comment]
Assuming ESPhome can be setup for an audio streaming source. I can think of 3 possible solutions we could work with.

1. The XMOS chip on the Sat1 Hat has 16 audio channel's of which we are using 6 (if I'm counting correctly) if we can convert line level audio to PDM we could read it in that way. The problem is we are running short of pins. But we could possibly read stereo in with 1 or 2 pins.

2. We could make a shim that adds a stereo line input to the I2S bus, that could then be used to read audio in, though this would require an additional chip for line level to I2S

(Alternatively if squeezelite works as an audio source, or ESPHome)
3. An external/separate device that contains a line level to I2S chip, and a ESP32. Theoretically this should be possible to be a rather simple circuit, and power over 5V usb

**Comment by Hedda** [comment]
To get more pins I guess that other additional solutions would maybe be possible if ever some day add a another ESP32 as as ansecond ”coprocessor” to the Satellite1, as suggested in the feature request linked bellow:

- https://github.com/FutureProofHomes/Satellite1-Hardware/issues/5

While it would make for a more complex solution you would then also have the option to off-load some processes to that second ESP32 as well if acting as a dedicared Thread Border Router or Bluetooth Proxy does not take up all of its reaources (which I believe it would not do if using the more powerful ESP32-C6 as that second ESP32)

**Comment by Hedda** [comment]
I assume that an ESP32-based solution for this also needs some kind of dedicated ADC (Analog-to-Digital-Converter) IC hardware chip specifically made for audio recording/capture that is purposely made for sound and communication via I2S or I2C protocols, so which existing chips on the market looks good for that and will those work fine with either ESPHome or XMOS firmware?

Also wondering if you guys think this concept should reuse the existing i2s microphone I²S Audio Microphone component for this purpose or if a new custom component for ESPHome would need to be created to specifically handle high-quality Line level stereo audio input to I2S for HiFi-quality music input from an analog source and digitalization of that?

- https://esphome.io/components/microphone/i2s_audio.html
  - https://esphome.io/components/microphone/i2s_audio.html#external-adc
  - https://esphome.io/components/microphone/i2s_audio.html#internal-adc

**UPDATE!** There is now also a new audio_adc component in ESPHome to allow a more standardized usage of Audio ADC hardware components:

- https://esphome.io/components/audio_adc/

PS: I have previously seen that [gnumpi](https://github.com/gnumpi), [rwrozelle](https://github.com/rwrozelle/audio-media-player), [nielsnl68](https://github.com/nielsnl68), and [luar123](https://github.com/luar123) have all been working on various advanced custom audio components for ESPHome that uses elements from the [Espressif Audio Development Framework (ADF)](https://github.com/espressif/esp-adf), and gnumpi's repository does mention ADC but it is way over my head.

- https://github.com/gnumpi/esphome_audio

- https://github.com/rwrozelle/audio-media-player

**Comment by alextrical** [comment]
Looking over the datasheets, the most likeley candidate chip to use may be a "PCM1808". That would allow Line level stereo to I2S.

**Comment by Hedda** [comment]
Tip! Found this article that compares some popular I2S ADC for Line-In audio feed:

* https://mm.kno.wled.ge/soundreactive/Line-Input/

That article includes all these chips:

- [TAC5112](https://www.ti.com/product/TAC5112) (new ADC and DAC combo IC from Texas Instruments)
- [TAA3020](https://www.ti.com/product/TAA3020)
- [PCM1863](https://www.ti.com/product/PCM1863) (an ADC which the [HiFiBerry DAC+ ADC Pro](https://www.hifiberry.com/docs/data-sheets/datasheet-dac2-adc-pro/) popular HAT/Sheild for Raspberry Pi uses in combination with [PCM5122](https://www.ti.com/product/PCM5122) DAC)
- [PCM1862](https://www.ti.com/product/PCM1862)
- [PCM1809](https://www.ti.com/product/PCM1809)
- [PCM1808](https://www.ti.com/product/PCM1808) (or [Ti PCM1802](https://www.ti.com/product/PCM1802))
- [TLV320ADC6120](https://www.ti.com/product/TLV320ADC6120)
- [TLV320ADC5120](https://www.ti.com/product/TLV320ADC5120)
- [TLV320ADC3120](https://www.ti.com/product/TLV320ADC3120)
- [TLV320AIC3204](https://www.ti.com/lit/gpn/TLV320AIC3204) (another ADC and DAC combo IC from Texas Instruments)
- [Everest Semiconductor ES7243E](https://www.lcsc.com/datasheet/lcsc_datasheet_2410121611_Everest-semi-Everest-Semiconductor-ES7243E_C2929446.pdf)
- [Everest Semiconductor ES8388](https://datasheet.lcsc.com/lcsc/1912111437_Everest-semi-Everest-Semiconductor-ES8388_C365736.pdf) and [AC101](https://files.seeedstudio.com/wiki/ReSpeaker_6-Mics_Circular_Array_kit_for_Raspberry_Pi/reg/AC101_User_Manual_v1.1.pdf)
- [Cirrus Logic WM8978](https://www.mouser.com/datasheet/2/76/WM8978_v4.5-1141768.pdf)
- [CirrusLogic WM8782](https://www.cirrus.com/products/wm8782/)
- [CirrusLogic CS5343](https://www.cirrus.com/products/cs5343-44/)
- [AKM AK5720](https://www.akm.com/eu/en/products/audio/audio-adc/ak5720et/) ([datasheet](https://www.akm.com/content/dam/documents/products/audio/audio-adc/ak5720vt/ak5720vt-en-datasheet.pdf))

Update: Perhaps best suited ADC IC alternative could be [TAC5112](https://www.ti.com/product/TAC5112) which has both ADC and DAC,? Or its predecessor [AIC3204 (TLV320AIC3204)](https://www.ti.com/product/TLV320AIC3204)?  (Noticed that while the TAC5112 is so new that ESPHome does not have any drivers for it, at least there is a DAC driver in ESPHome for AIC3204).

* https://www.ti.com/product/TAC5112

  * https://www.ti.com/lit/gpn/tac5112

    * https://www.ti.com/product/TLV320AIC3204/part-details/TLV320AIC3204IRHBR
    * https://www.ti.com/product/TLV320AIC3204/part-details/TLV320AIC3204IRHBT

Check out this "TAC5112EVM-K" complete evaluation kit (dvelopment board) for the TAC5112 but that cost a lot more than the IC alone:

https://www.ti.com/tool/TAC5112EVM-K

Note by the way that there is an a standard "TAC5112" and a matching Automotive SKU "TAC5112-Q1" so no idea which would be better(?).

FYI, it by the way looks like the "Muse Luxe Speaker" from Raspiaudio also has a 3.5mm headphone jack for "AUX" line level ADC stereo input (with their specifications saying up to 24-bit and 96 kHz sampling frequency) but again not sure if their existing ESPHome code that could make use of it for this purpose?

* https://raspiaudio.com/muse/

  * https://raspiaudio.com/product/esp-muse-luxe/

From what found it is either using ES8388 or WM8978 as external audio codec for the AUX jack? Looks like it has both, one for mic and one for the AUX jack? Seems like those can be used for either ADC or DAC or both?

* https://github.com/RASPIAUDIO/esphomeLuxe/tree/main

* https://github.com/RASPIAUDIO/Muse_library

* https://github.com/RASPIAUDIO/LuxeMicrophones

* https://github.com/RASPIAUDIO/es8388

* https://github.com/RASPIAUDIO/Simple_Bluetooth_Speaker_ESP32/tree/master

* http://www.everest-semi.com/pdf/ES8388%20DS.pdf

* https://www.mouser.com/datasheet/2/76/WM8978_v4.5-1141768.pdf?srsltid=AfmBOoqa0zkVrQbsG2qGh1gh7U-12vtnCpgCa6nd1mmNoMDA1XoVsnIX

Interestingly they also specifically mention this feature and the fact that its firmware is hackable in their marketing material.

![Image](https://github.com/user-attachments/assets/3422e9f2-b858-4c62-aa0b-60f6f14a8de0)

![Image](https://github.com/user-attachments/assets/2a166cd2-3a99-47b2-8a74-2fcb5ea28039)

PS: Off-topic but sounds as if that "Muse Luxe Speaker" is using the MAX98357A I2S Class D Amplifier DAC for output.

**Comment by that1guy** [comment]
@Hedda We do not have this feature locked in our roadmap, but we're open-minded to building it in the future once we get our current roadmap moving forward.  In, fact we like the idea of building a optional expansion model that can attach the Sat1 HAT which gives audio inputs.

I'd like to recommend you move this tickets thoughts to [a Github Discussion](https://github.com/orgs/FutureProofHomes/discussions) so people can upvote and help us focus on priorities that go beyond the current roadmap.  Thanks very much!

**Comment by Hedda** [comment]
> I'd like to recommend you move this tickets thoughts to [a Github Discussion](https://github.com/orgs/FutureProofHomes/discussions) so people can upvote and help us focus on priorities that go beyond the current roadmap. Thanks very much!

@that1guy only you as repository owners/admin have the rights/permissions to move tickets (we as non-members have no such permissions/rights).

Byw, you accidently closed this instead without meaning to and I do not have permissions to re-open it either.

**Reply by that1guy** [reply]
Converted to discussion.

**Comment by Hedda** [comment]
FYI, looks like ESPHome core develpers added Audio ADC core component to ESPHome 2025.2 version release:

https://esphome.io/components/audio_adc/

I believe that this new audio_adc component would allow your ESPHome devices to use audio ADC hardware components, and by extension allso allowing the sound capture / audio recording via the microcontroller from a range of sources, but need some kind of ADC chip for audio.

So far that Audio ADC core component support has support for [ES7210](https://esphome.io/components/audio_adc/es7210) and [ES7243E](https://esphome.io/components/audio_adc/es7243e) platforms.

https://esphome.io/changelog/2025.2.0.html#audio-components

**Reply by that1guy** [reply]
Great tip to share!  Thanks @Hedda.

**Comment by Hedda** [comment]
This "AUX line-in" feature is by the way a feature that Sonos offers on some of their products, either built-in or via a line-in adapter:

* https://support.sonos.com/en-us/article/use-line-in-on-sonos

* https://www.sonos.com/en-us/shop/sonos-line-in-adapter

* https://support.sonos.com/en-us/article/adjust-line-in-settings

Adapter can be used with different speakers but Sonos also has a dedicated "Sonos Port" Network Audio Streamer product with this feature:

* https://www.sonos.com/en-us/shop/port

![image](https://github.com/user-attachments/assets/39cdf406-6618-4d75-93ae-d60ec348179c)

PS: I also believe Sonos offer a similar feature on their soundbar for televisions as they allow users to stream audio from a TV to other speakers.

**Comment by Hedda** [comment]
@that1guy @gnumpi any chance you or other ESPHome developers you know could have an interest in helping out @alextrical with ESPHome software-side for his proof-of-concept ESP32-based  development board with ADC (analog-to-digital conversion) module and audio input ports?

* https://github.com/orgs/music-assistant/discussions/3763

ESPHome based Line-Level Audio ADC analogue-input for local-network multi-room music streaming:

  * https://github.com/alextrical/ESPHome-LineLevelADC

![image](https://github.com/user-attachments/assets/ca795736-ad43-4590-a4ad-9c22c932d5fe)

Quote:

_**Originally posted by @alextrical in https://github.com/music-assistant/support/discussions/3763**_
> _I am looking to help design a solution that will help work towards the request(s) #2343 for a Stereo line level audio input as a source to MusicAssistant, something that is embedded and doesn't rely on a RPi/Embedded linux device, while also allowing for a Remote 3.5mm AUX input jack_
> 
> I will get a Schematic drafted up, with the intent to make a open source PCB that can connect some ESP32-S3 and PCM1808 modules together to enable a 3.5mm jack to receive audio into an ESPHome based firmware.
> 
> _I will investigate and try to develop the Driver to add support for the PCM1808 in ESPHome if required._
> 
> _If this is something you would be willing to investigate, I will happily get the initial KiCad project setup, and repo made available on my Github. I can send you a hardware development board once ive tested and confirmed that the ESPHome side is functional_

_**Originally posted by @alextrical in https://github.com/music-assistant/support/discussions/3763#discussioncomment-12592036**_
> _This feature request is basically trying to work out what would be required for a 'Local Network' ESP32-S3 audio source to be able to stream audio to Music Assistant, and then forward that to the speakers in Music Assistant (basically treating the ESP32-S3 as a local network radio station)_

> _So the flow I would expect:_
> 
> * _**LP player - preamp - ESP32 with I2S line level ADC - WiFi - Music assistant - WiFi - Squeezelite player - Speakers**_

_**Originally posted by @alextrical in https://github.com/music-assistant/support/discussions/3763#discussioncomment-12586167**_
> _I believe solution that will likely solve the issue for @hedda and likely others, is a Remote WiFi audio input device. I know for certain that my media center, where I would want such a WiFi input sound card, is nowhere near to my server hosting Home Assistant._

_**Originally posted by @alextrical in https://github.com/music-assistant/support/discussions/2343#discussioncomment-12584820**_
> _The intent I am working towards, and that I think is a (elegant?) soultion to Remote 3.5mm is a ESP32-S3 ESHome based device.
I believe this will allow for a Low cost, stable device, that doesn't use a SBC/Embedded linux as its core.
This is currently a personal project as i see a need for it for others, but I believe it's something that FutureProofHomes would pick up and make a polished manufactured device available._

_**Originally posted by @alextrical in https://github.com/music-assistant/support/discussions/2343#discussioncomment-12582543**_
> _I've ordered a module to test the PCM1808 with  ESPHome, with the intent to verify it will work with the driver I will try and write._
> 
> _Once I get it working with ESPHome, are there any developers of Music Assistant I could send a hardware setup to, to integrate the functionality of a ESPHome audio input/source_.

_**Originally posted by @alextrical in https://github.com/music-assistant/support/discussions/2343#discussioncomment-12581584**_
> _This is likely the component in ESPHome I would guess we would use. I will look at getting a driver written up for the PCM1808_
> 
> https://github.com/esphome/esphome-docs/blob/current/components/microphone/i2s_audio.rst
> 
> _The question is how can we get that audio input piped into music assistant as an audio source?_


PS Again, this request for a more user-friendly ESP32 + ESPHome appliance plug-and-play solution is partially related to this other discussion:

* https://github.com/orgs/music-assistant/discussions/2343

**Reply by Hedda** [reply]
@that1guy @gnumpi Suggest also checkout HiFiBerry DAC+ ADC Pro (HAT for Raspberry Pi) datasheet/design with headers for add-ons:

* https://www.hifiberry.com/docs/data-sheets/datasheet-dac2-adc-pro/

  * https://www.hifiberry.com/shop/boards/dacplus-adc/

(HiFiBerry sells several compatible addon boards to addon different ports as well as amplifier board and even an DSP add-on board).

* https://hifiberry.com/shop/#accessories

Would be awesome with similar stand-alone product that uses onboard ESP32, and sound like kind of what alextrical will be prototyping:

* https://github.com/orgs/music-assistant/discussions/3763

Note: HiFiBerry DAC+ ADC Pro uses in [PCM1863](https://www.ti.com/product/PCM1863) IC for Audio ADC input, which it uses in combination with [PCM5122](https://www.ti.com/product/PCM5122) for DAC output.

PS: Suggest also look at new [TAC5112](https://www.ti.com/product/TAC5112) which is a ADC and DAC combo IC from Texas Instruments (so is maybe more cost-effective?).
